src/EFCore.PG/Query/Internal/NpgsqlSqlTranslatingExpressionVisitor.cs

```diff
@@ -383,5 +475,24 @@ bool TranslationFailed(Expression original, Expression translation, out SqlExpre
         }
 
         #endregion Copied from RelationalSqlTranslatingExpressionVisitor
+
+        public class ExpressionTypeExtractingExpressionVisitor<T> : ExpressionVisitor
```

Comment:
This type of visitors are bug prone.

Reply:
Can you provide more detail on why? The basic need here is for the configured case-sensitive/insensitive collation to bubble up the expression tree (a bit like how type mappings bubble up), so that we can specify it explicitly at the comparison level.

Reply:
TypeMapping used to have similar bubbling up code in past (albeit via a recursive method). But what does bubbling should do?
In expression `a + b` if `a` has collation and `b` does not then what is the collation of resulting expression?
Places where things are complicated
Suppose there is some function `Concat(a,b,c)`
what is resulting collation? 
- If only a has collation or only b has collation.
- If a,b,c all has collation but same value
- If a,b,c all has collation but different value
Further for translation of expression pattern like `"abc".IndexOf("b").ToString()`
IndexOf returns an int which is converted to string. It should not have barring with collation of "abc" or "b" but visitor would bubble it up still.

Reply:
The rule currently implemented, is that we expect exactly one collation to bubble out. If an expression (recursively) contains more than one column, and there are non-identical collations defined on those columns, we error ([see here](https://github.com/npgsql/efcore.pg/pull/1329/files#diff-c61aff25e85385e445833ca8538ef4acR234)). If there are no collations, we error as well (as today). This seems like a reasonable thing to do - do you disagree?

> Further for translation of expression pattern like "abc".IndexOf("b").ToString()
> IndexOf returns an int which is converted to string. It should not have barring with collation of "abc" or "b" but visitor would bubble it up still.

That's a good point. We could narrow things a bit, and say that we only bubble up the collation across operators/functions which also return a string (hence Concat would bubble up, but IndexOf wouldn't). I'm sure it's possible to find an example where this would be weird, but is there a better alternative? If we think this is too problematic, we could cut out translation of string.Equals with StringComparison, and require the user to explicitly define collations via EF.Functions.Collate - but there seems to be value in translating string.Equals and I'm not sure the actual problem is serious. What do you think?

PS this discussion seems very relevant also for SQL Server and MySQL support.

Reply:
Of course, one more possibility is only to look for a collation on the immediate parameters of string.Equals(). This would limit the usefulness a bit, but sounds like it would cover what most people actually need.

@smitpatel and others, thoughts?

Reply:
It is an approximate solution. We could put guards in to reduce false cases like only expecting one collation to bubble out. We could also improve correctness by blocking (or bubbling up) of certain operations. Or we could just not do it. At the end it would depend on how much cases it solves and what is the accuracy.
For typemapping it just did not work well for us. It had too many negative cases. At the same time, type mapping is a lot more widespread compared to this where you need only when translating string.Equals with Comparison. So it is not entirely bad. Also depends how much people are going to do complex operation before doing this kind of comparison. Most comparisons would be column to column/constant/parameter match only.
EF.Functions.Collate would always be an escape hatch for any implementation we add here (even if we don't try to find any expression this way and just use collation from CollateExpression generated by the function translation).
On two extremes
- We could keep this as is. May cause incorrect matches. Use escape hatch. Not sure how easy would it to identify incorrect match.
- We could remove this visitor, match probably directly CollateExpression or direct ColumnExpression. This should cover most basic cases. Relative use of basic case vs complex case is unknown to me.

An in between solution would be to let it go through very specific operations we know are safe. Like concat `string.Equals(a + b, c, ...)` Though we would need to define what would be behavior when only one of a/b has collation or both have different values. (Could also consider ScalarSubqueryExpression in this. Essentially form of `c.Orders.Select(...).FirstOrDefault()`). May be few more?
In any case where it did not work, translation would fail and user can use Collate API.

Note: When looking at collation, just like typeMapping, ignore the facet for constant/parameter. So a + constant should have same collation as a.